src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -262,31 +274,57 @@ class ExtentPlacementManager {
 
     // XXX: bp might be extended to point to different memory (e.g. PMem)
     // according to the allocator.
-    auto bp = ceph::bufferptr(
-      buffer::create_page_aligned(length));
-    bp.zero();
-    paddr_t addr;
+    std::list<alloc_result_t> allocs;
 #ifdef UNIT_TESTS_BUILT
     if (unlikely(external_paddr.has_value())) {
       assert(external_paddr->is_fake());
-      addr = *external_paddr;
+      auto bp = ceph::bufferptr(
+        buffer::create_page_aligned(length));
+      bp.zero();
+      allocs.emplace_back(alloc_result_t{*external_paddr, std::move(bp), gen});
     } else if (gen == INLINE_GENERATION) {
 #else
     if (gen == INLINE_GENERATION) {
 #endif
-      addr = make_record_relative_paddr(0);
+      auto addr = make_record_relative_paddr(0);
+      auto bp = ceph::bufferptr(
+        buffer::create_page_aligned(length));
+      bp.zero();
+      allocs.emplace_back(alloc_result_t{addr, std::move(bp), gen});
     } else if (category == data_category_t::DATA) {
       assert(data_writers_by_gen[generation_to_writer(gen)]);
-      addr = data_writers_by_gen[
+      auto addrs = data_writers_by_gen[
 	  generation_to_writer(gen)]->alloc_paddr(length);
```

Comment:
I have a fundamental question. If I remember correctly, there is a constraint that seastore need to allocate a consecutive address when allocating a block, right? From that perspective, It seems like `alloc_new_extents` invokes `alloc_paddr` with the desired length. This leads to `alloc_new_extents` could potentially return scattered blocks despite intending to acquire consecutive addresses. Am I missing something?

Reply:
> I have a fundamental question. If I remember correctly, there is a constraint that seastore need to allocate a consecutive address when allocating a block, right? From that perspective, It seems like `alloc_new_extents` invokes `alloc_paddr` with the desired length. This leads to `alloc_new_extents` could potentially return scattered blocks despite intending to acquire consecutive addresses. Am I missing something?

I think chances are that there can be a time when the disk space is very fragmented and that constraint can not be held. I think the bottom line is that we should allow IO requests to continue as long as there is enough disk space, no matter whether the remaining disk space is consecutive. It's true that we should always allocate consecutive space as long as it's possible; but if not, I think we should allow non-consecutive allocations; otherwise, we'll see failed allocations when there's still enough space, which I think is wrong. What do you think?

Reply:
> . If I remember correctly, there is a constraint that seastore need to allocate a consecutive address when allocating a block, right?

It seems that SeaStore doesn't have this constraint. Because Classic OSDs' ObjectStore doesn't have this constraint.

Reply:
@myoungwon object data resides in the logical address space and is consecutive there. So I think it should be fine that `alloc_new_extents()` returns non-consecutive physical addresses for data blocks.

Reply:
> I think, in RBM scenarios, chances are that there can be a time when the disk space is so fragmented that the constraint can not be held. I think the bottom line is that we should allow IO requests to continue as long as there is enough disk space, no matter whether the remaining disk space is consecutive. It's true that we should always allocate consecutive space as long as it's possible; but if not, I think we should allow non-consecutive allocations; otherwise, we'll see failed allocations when there's still enough space, which I think is wrong. Basically, I think we should drop the constraint if there was one, What do you think?

Yeah, the direction is the right way to go. But, I wonder if there are any potential risks to the system's reliability.

> @myoungwon object data resides in the logical address space and is consecutive there. So I think it should be fine that alloc_new_extents() returns non-consecutive physical addresses for data blocks.

This PR seems like the multiple allocation is used for not only data blocks but also metadata blocks. Is this ok? No worries if so.

Reply:
> This PR seems like the multiple allocation is used for not only data blocks but also metadata blocks. Is this ok? If so, I have no further concerns.

I've just modified the code so that metadata blocks are not affected, please take a look, again.

Reply:
Just wanted to point out that your code modification, which doesn't currently change behavior, could potentially have an effect in the future due to the use of `ceph_assert(extents.size() == 1);`  https://github.com/ceph/ceph/pull/54280#discussion_r1419898441 is similar to what I tried to say.